### ISSUE TITLE
Error when creating second alembic migration

### ISSUE THREAD

**Issue by Rubyj** [koxudaxi/py-data-api#35]
After successfuly creating an initial database with `alembic` and `sqlalchemy` I am having trouble adding a column in a second migration.

When I try to create the second migration (the database already exists from the first migration) I get the following error:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'variant_annotation_id_seq' as owned by integer column 'variant_annotation(id)', assuming SERIAL and omitting
Traceback (most recent call last):
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/config.py", line 573, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/config.py", line 567, in main
    self.run_cmd(cfg, options)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/config.py", line 547, in run_cmd
    **dict((k, getattr(options, k, None)) for k in kwarg)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/command.py", line 214, in revision
    script_directory.run_env()
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/script/base.py", line 489, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/util/compat.py", line 173, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "alembic/env.py", line 85, in <module>
    run_migrations_online()
  File "alembic/env.py", line 79, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/runtime/environment.py", line 846, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/runtime/migration.py", line 507, in run_migrations
    for step in self._migrations_fn(heads, self):
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/command.py", line 190, in retrieve_migrations
    revision_context.run_autogenerate(rev, context)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/api.py", line 433, in run_autogenerate
    self._run_environment(rev, migration_context, True)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/api.py", line 473, in _run_environment
    autogen_context, migration_script
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/compare.py", line 25, in _populate_migration_script
    _produce_net_changes(autogen_context, upgrade_ops)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/compare.py", line 51, in _produce_net_changes
    autogen_context, upgrade_ops, schemas
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/util/langhelpers.py", line 303, in go
    fn(*arg, **kw)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/compare.py", line 83, in _autogen_for_tables
    autogen_context,
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/alembic/autogenerate/compare.py", line 195, in _compare_tables
    inspector.reflecttable(t, None)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 667, in reflecttable
    reflection_options,
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 846, in _reflect_fk
    **reflection_options
  File "<string>", line 2, in __new__
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/util/deprecations.py", line 128, in warned
    return fn(*args, **kwargs)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 496, in __new__
    metadata._remove_table(name, schema)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
    raise value
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 491, in __new__
    table._init(name, metadata, *args, **kw)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 585, in _init
    resolve_fks=resolve_fks,
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 609, in _autoload
    _extend_on=_extend_on,
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1604, in run_callable
    return callable_(self, *args, **kwargs)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 431, in reflecttable
    table, include_columns, exclude_columns, resolve_fks, **opts
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 687, in reflecttable
    reflection_options,
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 942, in _reflect_unique_constraints
    constraints = self.get_unique_constraints(table_name, schema)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 520, in get_unique_constraints
    self.bind, table_name, schema, info_cache=self.info_cache, **kw
  File "<string>", line 2, in get_unique_constraints
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 56, in cache
    ret = fn(self, con, *args, **kw)
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3401, in get_unique_constraints
    for name, uc in uniques.items()
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3401, in <listcomp>
    for name, uc in uniques.items()
  File "/c/Users/rjacobs/git/vcf-parser-lambda/venv/lib/python3.6/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3400, in <listcomp>
    {"name": name, "column_names": [uc["cols"][i] for i in uc["key"]]}
KeyError: 'longValues'
```

Any idea @koxudaxi ? It seems like the unique id constraint is causing issues and preventing me from making second migration.

**Comment by koxudaxi** [comment]
@Rubyj 
I'm sorry for the late response.
Interesting. I don't have any idea. 
I see `KeyError: 'longValues'` on your error.
The library may not handle the value types.
Would you give me the table model of SQLAlchemy?
I will check it tonight or tomorrow.

**Comment by Rubyj** [comment]
@koxudaxi Thank you for responding!

I believe it has something to do with the sequence that is created for a primary key. That sequence seems to interfere when I try to modify the model.

Here are my models:

```
class FactoryMixin:
    @classmethod
    def map_to_model(cls, mapping: dict) -> 'FactoryMixin':
        mappings = inspect(cls).columns.keys()
        mappings.remove('id')
        mapping = {k: mapping[k] for k in mappings}
        return cls(**mapping)


class Variant(Base, FactoryMixin):
    __tablename__: str = 'variant'
    id = Column(Integer, primary_key=True)
    chrom = Column(String(25))
    pos = Column(Integer)
    rs_id = Column(String(100))
    ref = Column(String(75))
    alt = Column(String(75))
    qual = Column(Float)
    __table_args__ = (UniqueConstraint('chrom', 'pos', 'ref', 'alt'),)


class VariantAnnotation(Base, FactoryMixin):
    __tablename__ = 'variant_annotation'
    id = Column(Integer, primary_key=True)
    variant_id = Column(Integer, ForeignKey('variant.id'))
    variant = relationship(Variant, backref='annotations')
    allele = Column(String(75))
    annotation = Column(String(200))
    annotation_impact = Column(String(20))
    gene_name = Column(String(50))
    gene_id = Column(String(50))
    hgvs_c = Column(String(200))
    hgvs_p = Column(String(200))
    reference = Column(String(200))
    specimen_id = Column(String(50))
```

Here is the resulting migration for these models:
```
def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('variant',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('chrom', sa.String(length=25), nullable=True),
    sa.Column('pos', sa.Integer(), nullable=True),
    sa.Column('rs_id', sa.String(length=100), nullable=True),
    sa.Column('ref', sa.String(length=75), nullable=True),
    sa.Column('alt', sa.String(length=75), nullable=True),
    sa.Column('qual', sa.Float(), nullable=True),
    sa.PrimaryKeyConstraint('id'),
    sa.UniqueConstraint('chrom', 'pos', 'ref', 'alt')
    )
    op.create_table('variant_annotation',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('variant_id', sa.Integer(), nullable=True),
    sa.Column('allele', sa.String(length=75), nullable=True),
    sa.Column('annotation', sa.String(length=200), nullable=True),
    sa.Column('annotation_impact', sa.String(length=20), nullable=True),
    sa.Column('gene_name', sa.String(length=50), nullable=True),
    sa.Column('gene_id', sa.String(length=50), nullable=True),
    sa.Column('hgvs_c', sa.String(length=200), nullable=True),
    sa.Column('hgvs_p', sa.String(length=200), nullable=True),
    sa.Column('reference', sa.String(length=200), nullable=True),
    sa.Column('specimen_id', sa.String(length=50), nullable=True),
    sa.ForeignKeyConstraint(['variant_id'], ['variant.id'], ),
    sa.PrimaryKeyConstraint('id'),
    sa.UniqueConstraint('variant_id', 'specimen_id', 'allele', 'annotation', 'gene_name', 'hgvs_c', 'hgvs_p', 'reference')
    )
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('variant_annotation')
    op.drop_table('variant')
    # ### end Alembic commands ###
```

If I try to make **any** modification to these models and then generate a new revision migration with `alembic revision --autogenerate -m "Message"`. That is when I receive the error.

**Comment by Rubyj** [comment]
Just for reference. I created the issue here as well:

https://github.com/sqlalchemy/alembic/issues/618

**Comment by koxudaxi** [comment]
@Rubyj 
I found it!!
Sorry, I have not known the `arrayValues`.
https://docs.aws.amazon.com/rdsdataservice/latest/APIReference/API_ArrayValue.html#rdsdtataservice-Type-ArrayValue-longValues

The library have not supported the types yet.
I guess the types added this summer, but I did not notice.

I undersntad it when I watch the log.
> 2019-11-07 12:24:58,702 DEBUG sqlalchemy.engine.base.Engine Row ('variant_chrom_pos_ref_alt_key', {'longValues': [2, 3, 5, 6]}, 2, 'chrom')

Thank you very much. 
I will impliment it tomorrow.

**Comment by Rubyj** [comment]
@koxudaxi I think I have tracked down the issue.

If you run the following query in amazon's query editor (make sure to substitute a table OID):

```
            SELECT
                cons.conname as name,
                cons.conkey as key,
                a.attnum as col_num,
                a.attname as col_name
            FROM
                pg_catalog.pg_constraint cons
                join pg_attribute a
                  on cons.conrelid = a.attrelid AND
                    a.attnum = ANY(cons.conkey)
            WHERE
                cons.conrelid = :table_oid AND
                cons.contype = 'u'
```

I see the following results:

![image](https://user-images.githubusercontent.com/4855908/68419038-7e460a00-0167-11ea-856e-363a5eb281ad.png)

This makes total sense. However, when I run this query in code using `engine.execute(statement)` with `statement` being this SQL and `engine` being my engine, I see the following results:

```
2019-11-07 14:02:15,831 DEBUG sqlalchemy.engine.base.Engine Row ('variant_chrom_pos_ref_alt_key', {'longValues': [2, 3, 5, 6]}, 2, 'chrom')

2019-11-07 14:02:15,832 DEBUG sqlalchemy.engine.base.Engine Row ('variant_chrom_pos_ref_alt_key', {'longValues': [2, 3, 5, 6]}, 3, 'pos')

2019-11-07 14:02:15,832 DEBUG sqlalchemy.engine.base.Engine Row ('variant_chrom_pos_ref_alt_key', {'longValues': [2, 3, 5, 6]}, 5, 'ref')

2019-11-07 14:02:15,833 DEBUG sqlalchemy.engine.base.Engine Row ('variant_chrom_pos_ref_alt_key', {'longValues': [2, 3, 5, 6]}, 6, 'alt')
```

It looks to me like the `key` column is an array, and for some reason when queried in the code, it prints out as a dictionary with a key of 'longValue'. Can we change the way array is being handled by pydataapi?

**Comment by koxudaxi** [comment]
@Rubyj 
Now, pydataapi returns raw value. 
We implement it to handle `arrayValues` in this line. 
https://github.com/koxudaxi/py-data-api/blob/0299e2474f927281c2bcadd9b31394f4de38daab/pydataapi/pydataapi.py#L243-L250

**Comment by Rubyj** [comment]
Sounds good. 

I imagine this is responsible as well.

https://github.com/koxudaxi/py-data-api/blob/master/pydataapi/pydataapi.py#L105

**Comment by koxudaxi** [comment]
The issue is closed because I have released it as version 0.4.8.
